Introduce DeepScan access token for VS Code

jihye1 opened the pull request:
- Create deepscanToken.ts to manage DeepScan access token
- Add 2 commands for a user to configure and delete access token
- Change DeepScan inspect API from /api/demo to /api/vscode/analysis
  - Add the access token into API request Authorization header
  - Handle inspect fail notifications when the token does not exist or is invalid, or has expired.
- Bump @types/vscode from 1.25.0 to 1.53.0 to use SecretStorage
  - Fix some type errors in StatusBar.ts
  - Change compilerOptions form es6 to es2018 in tsconfig.json

jihye1 committed:
Introduce DeepScan access token for VS Code

- Create deepscanToken.ts to manage DeepScan access token
- Add 2 commands for a user to configure and delete access token
- Change DeepScan inspect API from /api/demo to /api/vscode/analysis
  - Add the access token into API request Authorization header
  - Handle inspect fail notifications when the token does not exist or
    is invalid, or has expired.
- Bump @types/vscode from 1.25.0 to 1.53.0 to use SecretStorage
  - Fix some type errors in StatusBar.ts
  - Change compilerOptions form es6 to es2018 in tsconfig.json

Signed-off-by: Jihye Won <jihye.won1@samsung.com>

jihye1 committed:
Change package version (1.62.1)

Signed-off-by: Jihye Won <jihye.won1@samsung.com>

ws807 commented on client/src/extension.ts:
```diff
@@ -285,6 +318,22 @@ function registerEmbeddedCommand(command: string, handler) {
     return embeddedCommand;
 }
 
+async function checkDeepscanToken(context, deepscanToken) {
```
인자 타입 기재해 주세요.

ws807 commented on client/src/extension.ts:
```diff
@@ -285,6 +318,22 @@ function registerEmbeddedCommand(command: string, handler) {
+async function checkDeepscanToken(context, deepscanToken) {
+    const config = getDeepScanConfiguration();
+
+    if (config.get('enable') === false || packageJSON.version !== '1.63.0' ) {
```
packageJSON.version !== '1.63.0' 조건은 삭제해 주세요.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+    });
+  }
+
+  async showOneOffTokenNotification() {
```
논의한대로 이름을 showActivationNotification() 으로 바꾸기로 하죠.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+    return selected;
+  }
+
+  async showEmptyTokenNotification() {
```
논의한대로 토큰이 없어서 분석 실패하는 경우에 대해선 inspect failure notification은 보여주지 말기로 하죠.

Activation notification 보여주지 않은 상황에서만 이 notification 보여주는 것으로 중복을 막기로 했었는데 다음의 이유로 아예 일괄적으로 보여주지 않는 것이 적절할 듯 합니다.

- 플러그인 업데이트 후 reload 전에는 예전 버전이 이용되므로 토큰 없더라도 분석 실패 발생하지 않음
- 플러그인 업데이트 후 reload 한다면 activation notification 보게 됨
- Activation notification에서 Don't show again 누른 사용자
  - 사용자 관점에서 activation notification과 inspect failure notification의 차이를 알기 어려움
  - 따라서 그냥 토큰이 없는 상황에 대한 메시지를 더 이상 보지 않는 것을 원한다고 봐야함
- 사용자가 토큰 delete 한 경우
  - Delete는 사용자의 명시적 action 이므로 이로 인해 발생한 에러에 대해선 notification 보여주지 않는 것이 적절
  - 아울러 VS Code 다시 시작하는 시점까지 다시 토큰 등록하지 않는다면 activation notification 보여지게 됨

이렇게 한다면 activation notification과 inspect failure notification을 완전히 분리해서 처리가 가능해집니다.
따라서 global store의 상태 변수는 isInspectFailureNotificationDisabled 라는 이름으로 두고 다음과 같이 관리하면 될 듯 합니다.

- 기본 설정은 false
- Invalid 또는 expired token notification 보여주기 직전에 true로 설정
  - VS Code가 알아서 동시에 동일 notification 여러 번 띄우지 않게 처리해주고 있기는 함
  - 하지만 notification 창 닫은 후에 save 할 때마다 다시 notification 뜨는 것은 너무 과한 수준임
  - 에러가 발생 상황은 status bar 등에서 계속 표시되므로 notification은 보조적인 수단으로 볼 수 있어 최소한으로 보여주는 것이 타당함
- 토큰 등록되면 false로 설정
  - 차후에 다시 invalid 혹은 expired 상태가 됐을 때는 notification 다시 보여줘야 함

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+  private readonly _secretStorage: vscode.SecretStorage;
+  private tokenName: string;
+  private serverUrl: string;
+  private tokenRenerateUrl: string;
```
```suggestion
  private tokenRegenerateUrl: string;
```

ws807 commented on client/src/extension.ts:
```diff
@@ -81,27 +101,27 @@ async function activateClient(context: vscode.ExtensionContext) {
         const isChanged = isConfigurationChanged('fileSuffixes', oldConfig, newConfig) ||
                           isConfigurationChanged('serverEmbedded.enable', oldConfig, newConfig) ||
                           isConfigurationChanged('serverEmbedded.serverJar', oldConfig, newConfig);
+        const isChangedIgnoreConfig = isConfigurationChanged('ignoreRules', oldConfig, newConfig) ||
+                          isConfigurationChanged('ignorePatterns', oldConfig, newConfig);
```
ignoreRules 및 ignorePatterns 변경 사항은 reload 전에 실시간으로 반영해 주지 않기로 하지 않았나요?

이제 맞춰서 server.ts의 onDidChangeConfiguration() callback에서 ignoreRules 및 ignorePatterns 변경 사항 반영하는 코드 제거해 주세요.

ws807 commented on server/src/server.ts:
```diff
@@ -257,11 +271,28 @@ function parseProxy(proxyUrl) {
+            message: null,
+            uri: null,
+        });
+        return;
```
return 전에 sendDiagnostics([]); 수행해서 기존 이슈 제거해 주세요.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+    vscode.commands.registerCommand('deepscan.setToken', async () => {
+      let tokenInput: string = await vscode.window.showInputBox({
+          title: "Configure DeepScan Access Token For VS Code"
+      }) ?? '';
```
사용자가 입력을 취소한 경우에 showInputBox() 반환 값은 undefined 입니다.
그런데 이를 empty string 입력한 것과 동일하게 취급하면 ESC 누른 경우에도 토큰 값이 틀렸다는 메시지 창이 보여지게 됩니다.
적절히 수정해 주세요.

jihye1 replied:
네,  validateInput 옵션을 사용하여 empty string 입력한 경우 "Access token cannot be blank." 보여지게 됩니다.

참고로, ESC 누른 경우에는 command 창이 사라집니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+          title: "Configure DeepScan Access Token For VS Code"
+      }) ?? '';
+      tokenInput = tokenInput.trim();
+      if (this._validateToken(tokenInput)) {
```
별도 메시지창보다는 showInputBox 옵션에 있는 validateInput 이용하기로 하죠.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+
+  private _registerCommands() {
+    vscode.commands.registerCommand('deepscan.setToken', async () => {
+      let tokenInput: string = await vscode.window.showInputBox({
```
- placeHolder 옵션 지정해 주세요.
  - 대략 'Paste your token here' 라고 하면 될것 같네요
- password 옵션 지정하기로 하죠.
  - 토큰값이 별 생각없이 그냥 이상하다고 느낄 사용자가 있을 것도 같습니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+        const deleteAction = 'Delete';
+        const cancel = 'Cancel';
+        const message = `Are you sure you want to delete your DeepScan access token? The DeepScan extension will no longer be able to inspect.`;
+        const selected = await vscode.window.showWarningMessage(message, deleteAction, cancel);
```
Delete 할지 말지 선택하는 것은 곧바로 취해야 하는 action 입니다.
따라서 modal로 보여주는 것이 좋을 듯 합니다.

참고로 modal 이용할 때는 action을 MessageItem 타입으로 지정하고 cancel action에는 isCloseAffordance 속성을 true 로 설정해야하네요.
이렇게 하지 않으면 default로 Cancel 버튼이 보여져서 Cancel 버튼이 두 개가 되는 상황이 발생합니다.
Cancel action을 아예 지정하지 않는 방안도 있겠지만 그렇게 하는 것은 왠지 안정적이지 않을 듯 하니 MessageItem 이용하기로 하죠.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+        const selected = await vscode.window.showWarningMessage(message, deleteAction, cancel);
+        if (selected === deleteAction) {
+          this.deleteToken();
+          vscode.window.showInformationMessage(`Deepscan access token is successfully deleted.`);
```
마찬가지로 notification 메시지 대신에 Output 창에 다음과 같이 출력해 주세요.

- DeepScan access token was deleted.

updateTokenRequest 의 sendRequest successCallback 에서 토큰 값이 undefined 인 경우 delete 라고 판단하면 될 것 같네요.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+      }) ?? '';
+      tokenInput = tokenInput.trim();
+      if (this._validateToken(tokenInput)) {
+        this.setToken(tokenInput);
```
SecretStorage.store()는 Thenable 반환하네요.
this.setToken()이 해당 Thenable 반환하도록 수정하고 여기에선 await 추가해 주세요.

아래의 this.deleteToken()도 유사하게 수정해 주고요.

ws807 commented on server/src/server.ts:
```diff
@@ -331,12 +363,26 @@ async function inspect(identifier: VersionedTextDocumentIdentifier) {
         // Clear problems
         sendDiagnostics([]);
+        let state: Status | TokenStatus = Status.fail;
+        if (message.includes('expired')) {
```
'expired' 혹은 'Invalid' 문구만 보고서 토큰 관련 에러하고 인식하는 것은 충분치 않은 것 같습니다.
'token' 이라는 문구를 추가로 체크하는게 어떨까요?

jihye1 replied:
네, 서버 error message 전체 문장을 체크하려했는데, 일단 두 문구로 체크하도록 하겠습니다.

ws807 commented on package.json:
```diff
@@ -125,6 +125,16 @@
+            {
+                "category": "DeepScan",
+                "command": "deepscan.setToken",
+                "title": "Configure DeepScan Access Token For VS Code"
```
커맨드 이름은 최대한 간단하게 하는 것이 좋을 듯 합니다.
DeepScan은 prefix로 붙으니 이름에 다시 넣을 필요 없을 듯 하고요.

```suggestion
                "title": "Configure Access Token"
```

ws807 commented on package.json:
```diff
@@ -125,6 +125,16 @@
+            {
+                "category": "DeepScan",
+                "command": "deepscan.deleteToken",
+                "title": "Delete DeepScan Access Token For VS Code"
```
```suggestion
                "title": "Delete Access Token"
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+  private _registerCommands() {
+    vscode.commands.registerCommand('deepscan.setToken', async () => {
+      let tokenInput: string = await vscode.window.showInputBox({
+          title: "Configure DeepScan Access Token For VS Code"
```
```suggestion
          title: "Configure DeepScan Access Token"
```

ws807 commented on client/src/types.ts:
```diff
@@ -17,6 +18,13 @@ export enum Status {
     fail = 3 // Analysis failed
 }
 
+export enum TokenStatus {
```
논의한대로 별도 타입으로 정의하지 말고 기존 Status 확장하기로 하죠.

그리고 Status 사용처 전부 확인해서 토큰 관련 사항 추가해 주고요.
일단 StatusBar.update() 에서 토큰 처리가 빠져 있어서 토큰 에러 발생 시 우측 하단의 "DeepScan" 문구의 색깔이 바뀌지 않고 있습니다.

ws807 commented on client/src/extension.ts:
```diff
@@ -208,9 +232,13 @@ async function activateClient(context: vscode.ExtensionContext) {
+                updateStatus(state as Status);
+                showNotificationIfNeeded(params);
+                activeDecorations.updateDecorations(uri);
+            }
```
토큰 에러인 경우에도 else 브랜치의 updateStatus() 등의 동작 전부 수행되도록 해주세요.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+
+  private _validateToken(token: string) {
+   if (!token) {
+      vscode.window.showWarningMessage(`Sorry, DeepScan access token for VS Code cannot be blank.`);
```
```suggestion
      vscode.window.showWarningMessage(`Access token cannot be blank.`);
```

"Sorry,"는 뺴주세요.
Sorry는 서버 다운, 알수 없는 서버 오류 등과 같이 우리측의 문제로 기능을 제공해주지 못할 때 쓰는 것이 적절한 것 같습니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+      if (token) {
+        const deleteAction = 'Delete';
+        const cancel = 'Cancel';
+        const message = `Are you sure you want to delete your DeepScan access token? The DeepScan extension will no longer be able to inspect.`;
```
```suggestion
        const message = `Are you sure you want to delete the DeepScan access token? DeepScan extension will no longer be able to inspect your code.`;
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+          vscode.window.showInformationMessage(`Deepscan access token is successfully deleted.`);
+        }
+      } else {
+        vscode.window.showInformationMessage(`Sorry, DeepScan access token doesn't exist.`);
```
```suggestion
        vscode.window.showInformationMessage(`Nothing to delete. DeepScan access token is currently not registered.`);
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+
+  async showInvalidTokenNotification() {
+    const check = 'Go to DeepScan';
+    const message = `Sorry, DeepScan access token is invalid. Check to make sure you copy the correct token string.`;
```
```suggestion
    const message = `Your DeepScan access token is not valid. Regenerate it and make sure to copy the currect token string.`;
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  async showExpiredTokenNotification() {
+    const regenerate = 'Regenerate';
```
버튼이 하나여서 썰렁하기도 하니 다음과 같이 수정하기로 하죠.

```suggestion
    const regenerate = 'Regenerate Token';
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  async showInvalidTokenNotification() {
+    const check = 'Go to DeepScan';
```
사용중인 token을 revoke한 경우가 가장 일반적으로 invalid 해지는 상황일 듯 합니다.
따라서 Expired와 동일하게 Regenerate Token 버튼 두기로 하죠.
(링크도 동일하게 account settings로 하면 될 듯 합니다.)

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+  async showExpiredTokenNotification() {
+    const regenerate = 'Regenerate';
+    const message = `Sorry, DeepScan access token has expired.
+    You can regenerate a new token by clicking on the button below to go to DeepScan Account Settings page.`;
```
```suggestion
    const message = `Your DeepScan access token has expired. Regenerate it to continue inspecting your code with DeepScan.`;
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+    context.subscriptions.push(context.secrets.onDidChange((e) => this._handleSecretChange(e)));
+    this.tokenName = 'deepscan-token';
+    this.serverUrl = serverUrl;
+    this.tokenRenerateUrl = `${serverUrl}/auth/git?redirect_uri=/dashboard/#view=account-settings`;
```
DeepScan에 가입한적 없거나 탈퇴한 후에 OAuth 권한을 revoke한 사용자의 경우 이 URL 이용 시 GitHub의 "Authorize DeepScan" 페이지가 보입니다.
그런데 authorize 하고 가입한 후에 account-settings로 이동하지 않고 그냥 dashboard가 보이네요.

이미 가입돼 있거나 탈퇴했어도 OAuth 권한 revoke 하지 않은 사용자의 경우에만account-settings로 이동하는듯 합니다.

그런데 DeepScan에 가입한적 없는 사용자에게 곧바로 "Authorize DeepScan" 페이지를 보여주는 것 자체가 좀 이상한 듯 합니다.
사용자는 그냥 토큰 생성한다고만 생각하고 authorize 했는데 덜컥 의도치 않게 DeepScan에 가입되버린 상황이라고 느낄 것도 같아서요.

그냥 "Generate" 버튼에 "https://deepscan.io/docs/deepscan/vscode#token" 페이지 링크를 다는 것이 적절할 듯 합니다.
너무 편리하게 곧바로 account settings로 이동시켜주기보다는 가입 후에 dashboard를 보여줘서 둘러보다가 프로젝트 분석도 수행해봐라는 "원래 의도"에 좀 더 부합되는 것 같고요.

jihye1 replied:
네, Generate 버튼은 guide document link 로 달고, Regenerate 버튼은 account-settings 페이지 link 를 다는 것으로 수정하였습니다.

ws807 commented on client/tsconfig.json:
```diff
@@ -5,7 +5,7 @@
         "rootDir": "src",
         "lib": [
-            "es6"
+            "es2018"
```
의도치 않게 @types/vscode 버전이 1.91.0로 돼서 es2018 필요했었던 것 같네요.
다시 "es6"로 되돌려 주세요.

ws807 commented on client/src/StatusBar.ts:
```diff
@@ -36,7 +36,7 @@ export class StatusBar {
     }
 
-    setTooltip(text: string) {
+    setTooltip(text: string | vscode.MarkdownString) {
```
StatusBarItem 클릭하면 Output 창으로 이동하고 거기서 DeepScan 링크보여지게 되니 tooltop에 Markdown 이용해서 링크 포함시킬 필요는 없을 듯 합니다.

Tooltip의 Markdown은 VS Code 1.59.0 버전부터 지원되는 기능이기도 하고요.

jihye1 replied:
네, 필요없었지만, @type/vscode 버전을 1.91로 올리면서 vscode.statusBarItem.tooltip 의 type을 vscode.MarkdownString | string 으로 정해줬어야 했어서 포함시켰습니다.
버전을 1.56.0 으로 내리면서, 다시 원복해두었습니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+      tokenInput = tokenInput.trim();
+      if (this._validateToken(tokenInput)) {
+        this.setToken(tokenInput);
+        vscode.window.showInformationMessage(`DeepScan access token has successfully configured.`);
```
동작시켜보니 Enter 누르자마자 곧바로 뜨는 메시지창은 조금 과한 느낌이네요.
그냥 LanguageClient#info() 이용해서 Ouput 창에 출력하기로 하죠.
메시지는 다음과 같이 해주고요.

- New DeepScan access token was configured.

jihye1 replied:
네, updateTokenRequest 함수 안에서 sendRequest 의 successCallback 에서 output 창에 출력하도록 수정하였습니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+    const neverShowAgain = 'Don\'t show again';
+    const message = `DeepScan access token is required for using the DeepScan extension.
+    The token was introduced as a step towards providing personalization features.
+    You can generate a new token by clicking on the button below to go to DeepScan Account Settings page.`;
```
```suggestion
    const message =
      'An access token is required for using the DeepScan extension. ' +
      'DeepScan server uses the token to provide reliable and managed inspection of your code.';
```

Multiline template literal 사용하면 newline 문자가 포함되게 되는데 차후에 VS Code에서 이를 보여주는 방식이 바뀔 수도 있을 듯 합니다.
따라서 그냥 space로 문장 구분하는 것이 좋을 듯 합니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+        return;
+    }
+
+    const generate = 'Generate';
```
```suggestion
    const generate = 'Generate Token';
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,122 @@
+    You can regenerate a new token by clicking on the button below to go to DeepScan Account Settings page.`;
+    const selected = await vscode.window.showErrorMessage(message, regenerate);
+    if (selected === regenerate) {
+      vscode.env.openExternal(vscode.Uri.parse(this.tokenRenerateUrl));
```
토큰이 expire한 경우는 이미 DeepScan에 가입했다고 볼 수 있으니 auth 거치지 않는 다음 링크 사용해 주세요.

- https://deepscan.io/dashboard/#view=account-settings

ws807 commented on server/src/server.ts:
```diff
@@ -257,11 +271,28 @@ function parseProxy(proxyUrl) {
+        + `    2. Back to VS Code and open the Command Palette\n`
+        + `    3. Type in 'Configure DeepScan Access Token For VS Code'\n`
+        + `    4. Paste in your newly generated token value.\n`
+        + `Please, visit DeepScan Guide(${guideUrl}) for more information.`);
```
```suggestion
        connection.console.error(`Failed to inspect: DeepScan access token not configured. Visit ${guideUrl} to generate a token.`);
```

그냥 guide 문서 링크하기로 하죠.
Output 창에 자세한 instruction을 기재하는 것은 아무래도 시인성이 떨어지네요.
그리고 분석 실패할때마나 보여질테니 간결한 필요가 있는 것 같고요.

ws807 commented on server/src/server.ts:
```diff
@@ -331,12 +363,26 @@ async function inspect(identifier: VersionedTextDocumentIdentifier) {
+        } else if (message.includes('Invalid')) {
+            state = TokenStatus.invalid;
+            message = `Sorry, DeepScan access token is invalid. Check to make sure you copy the correct token value.\n`
+            + `Please, visit DeepScan Guide(${guideUrl}) for more information.`;
```
```suggestion
            message = `invalid DeepScan access token. Visit ${generateUrl} to regenerate the token and make sure to copy the correct token string.`;
```

ws807 commented on server/src/server.ts:
```diff
@@ -331,12 +363,26 @@ async function inspect(identifier: VersionedTextDocumentIdentifier) {
+            + `    2. Back to VS Code and open the Command Palette\n`
+            + `    3. Type in 'Configure DeepScan Access Token For VS Code'\n`
+            + `    4. Paste in your newly generated token value.\n`
+            + `Please, visit DeepScan Guide(${guideUrl}) for more information.`;
```
```suggestion
            message = `DeepScan access token expired. Visit ${generateUrl} to regenerate the token.`;
```

ws807 commented on client/package.json:
```diff
@@ -15,7 +15,7 @@
     "scripts": {},
     "dependencies": {
-        "@types/vscode": "^1.25.0",
+        "@types/vscode": "^1.53.0",
```
```suggestion
        "@types/vscode": "~1.56.0",
```

package-lock.json을 보니 1.91.0이 설치되었네요.
"^" 대신에 "~" 사용해 주세요.

그리고 InputBoxOptions.title은 1.56.0 버전부터 지원되고 있네요.

engines.vscode 속성은 "^1.56.0"으로 수정해 주고요.

jihye1 replied:
네, engines.vscode 속성을 포함하여, client/package-lock.json 파일도 npm i 로 1.91 -> 1.56 으로 수정하였습니다.

ws807 commented on client/src/extension.ts:
```diff
@@ -325,7 +374,11 @@ function getDeepScanConfiguration(): vscode.WorkspaceConfiguration {
+}
+
+function getServerUrl(): string {
+    return detachSlash(getDeepScanConfiguration().get('server')) || 'https://deepscan.io';
```
코드 다른 부분을 보면 getDeepScanConfiguration() 반환값에 대해 truthy 체크를 사용하고 있네요.
언제 null일 수 있는지는 모르겠지만 여기서도 안전하게 체크하는 것이 좋을 듯 합니다.

ws807 commented on client/src/utils.ts:
```diff
@@ -25,3 +25,24 @@ export function sendRequest(client: LanguageClient, command: string, successCall
+    });
+}
+
+export function detachSlash(path) {
```
인자 및 반환 타입 기재헤 주세요.

ws807 commented on package.json:
```diff
@@ -12,7 +12,7 @@
     "icon": "deepscan_icon.png",
     "engines": {
-        "vscode": "^1.25.0"
+        "vscode": "^1.53.0"
```
```suggestion
        "vscode": "^1.56.0"
```

client/package.json 에도 engines.vscode 속성 있는데 같이 수정해 주세요.

ws807 reviewed (changes requested):
큰 상관은 없는 것 같기는 한데 server/package.json 에서 버전을 1.5.0 -> 1.6.0 으로 올려 주세요.

jihye1 committed:
Apply suggestions from code review

Co-authored-by: ws807 <41432352+ws807@users.noreply.github.com>

jihye1 committed:
client, server: Apply suggestions from code review

- Modify code not to show inspect failure notifications
  if an access token is not provided.
- UX changes to status bar, notifications, and console logs
- Fix @type/vscode version (^1.56.0 -> ~1.56.0)

Signed-off-by: Jihye Won <jihye.won1@samsung.com>

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,112 @@
+          }
+      });
+      if (tokenInput) {
+        await this.setToken(tokenInput);
```
```suggestion
        await this.setToken(tokenInput.trim());
```

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,112 @@
+        const selected = await vscode.window.showWarningMessage(message, { modal: true }, deleteAction, cancelAction);
+        if (selected === deleteAction) {
+          await this.deleteToken();
+          vscode.window.showInformationMessage(`Deepscan access token is successfully deleted.`);
```
Delete 하는 경우에도 별도 notofication이 아니라 Output 창에 다음과 같이 출력되게 해주세요.

* DeepScan access token was deleted.

updateTokenRequest 의 sendRequest successCallback 에서 토큰 값이 undefined 인 경우 delete 라고 판단하면 될 것 같네요.

jihye1 replied:
네, 수정하였습니다. 이전 commit에서 resolved 하지 않았는데, 제가 resolved를 눌렀나보네요;

ws807 commented on client/src/extension.ts:
토큰 에러 발생한 경우에 Output 창과 동일한 메시지를 status bar message로 보여주는 것은 너무 긴듯 합니다.
기존 서버 에러 형식에 맞춰서 다음과 같이 간략하게 표현하기로 하죠.
(showNotificationIfNeeded() 함수 참고)

* A problem occurred communicating with DeepScan server. (no access token)
* A problem occurred communicating with DeepScan server. (invalid access token)
* A problem occurred communicating with DeepScan server. (expired access token)

jihye1 replied:
네, showNotificationIfNeeded 안에서 params.state 에 따라 status bar message 처리하였습니다.

ws807 commented on client/src/extension.ts:
```diff
@@ -285,6 +314,21 @@ function registerEmbeddedCommand(command: string, handler) {
+    if (config.get('enable') === false) {
+        return;
+    }
+    if (context.globalState.get('isInspectFailureNotificationDisabled') === true) {
```
Activation notification의 "Don't show again" 버튼을 누른 것과 isInspectFailureNotificationDisabled 플래그는 별개 조건입니다.

"Don't show again"에서는 기존대로 ignoreTokenWarning 플래그를 쓰고  isInspectFailureNotificationDisabled 는 별개로 처리해야 합니다.
그런데 이름 때문에 두 플래그의 역할이 혼동이 되는 것 같네요.

다음과 같은 이름으로 두 개의 플래그 사용하도록 수정하기로 하죠.

* isEmptyTokenWarningDisabled
  - Extension activation 시에 empty token 에 대해 경고창 보여줄지를 결정
  - 사용자가 "Don't show again" 누르면 true로 설정
* isExpiredOrInvalidTokenWarningDisabled
  - 매 분석 실패할 때마다 invalid/expired token 에 대해 경고창 보여줄지를 결정
  - 한 번 경고창 보여주면 곧바로 true로 설정
  - 파일 저장할 때마다 계속 경고창이 뜨는 것을 막으려는 용도
    - 현재는 경고창 뜨고 이를 "x"로 닫고 다시 파일 저장하면 계속 뜹니다.
  - 새로운 토큰 값 설정하면 다시 false로 설정
    -  새로 설정한 토큰이 추후 다시 expire 될때 경고창 보여주기 위해 필요

jihye1 replied:
네... 너무 헷갈렸는데 정리해주셨네요.. 위와 같이 수정하였습니다.

ws807 commented on client/src/extension.ts:
```diff
@@ -140,6 +157,9 @@ async function activateClient(context: vscode.ExtensionContext) {
     let activeDecorations;
 
+    const serverUrl = getServerUrl();
+    const deepscanToken = new DeepscanToken(context, serverUrl);
```
Server 주소가 바뀐 경우 해당 주소가 DeepscanToken 객체에 적용이 되지 않고 있네요.

Server 주소는 토큰 에러에 대한 메시지 창 보여줄 때만 필요하므로 DeepscanToken 객체 생성 시점에 미리 계산해 놓는 것보다 에러 발생했을때 configuration에서 읽어오도록 하는 것이 어떨까요?

jihye1 replied:
네, 미처 생각하지 못한 부분이네요. 반영하였습니다.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  private async _handleSecretChange(e: vscode.SecretStorageChangeEvent) {
+    vscode.commands.executeCommand('deepscan.updateToken');
```
e.key가 this.tokenName과 같을때만 executeCommand 수행해 주세요.

ws807 commented on package.json:
deepscan.ignoreRules 및 deepscan.ignorePatterns 설명에 "(Restart required)" 문구 추가해 주세요.

ws807 commented on client/src/deepscanToken.ts:
```diff
@@ -0,0 +1,112 @@
+    const neverShowAgain = 'Don\'t show again';
+    const message =
+      'An access token is required for using the DeepScan extension. ' +
+      'DeepScan server uses the token to provide reliable and managed inspection of your code.';
```
```suggestion
      'DeepScan server uses the token to provide reliable and prompt inspection of your code.';
```

jihye1 committed:
client, server: Handle token warning notifications with flags

- Handle token warning notifications with flags
- Modify status bar message when the token is not provided, invalid or
expired.
- Remove serverUrl property in DeepScanToken because serverUrl
configuration can be changed dynamically

ws807 commented on client/src/extension.ts:
```diff
@@ -53,17 +53,18 @@ async function activateClient(context: vscode.ExtensionContext) {
+                await deepscanToken.showInvalidTokenNotification(getServerUrl());
                 break;
         }
+        context.globalState.update('isExpiredOrInvalidTokenWarningDisabled', true);
```
큰 상관은 없는 것 같기는 한데 switch() 구문 이전으로 옮겨서 경고창 보여주기 직전에 true로 설정하는 것이 좋을 듯 합니다.

jihye1 replied:
네, 간단해서 수정 commit 바로 올렸습니다.

ws807 committed:
Add access token documentation at README.md

jihye1 committed:
client: Update globalState flag before showing a token notification

ws807 committed:
Add "Show Access Token Info" command

- Etc: adjust some token error messages

ws807 committed:
Adjustment for embedded mode

- Do not show empty token warning at activation in embedded mode
- Disable token commands in embedded mode
